src/pybind/mgr/mgr_module.py

```diff
@@ -326,6 +390,32 @@ def __del__(self):
     def log(self):
         return self._logger
 
+    @classmethod
+    def _traverse_module_packages(cls, root, package):
+        mods = [mod for _, mod, _ in pkgutil.iter_modules([root])]
+        for mod_name in mods:
+            try:
+                mod = importlib.import_module('.{}'.format(mod_name),
+                                              package=package)
+                if '__package__' in mod.__dict__:
+                    if mod.__dict__['__package__'] != package:
+                        cls._traverse_module_packages("{}/{}".format(root, mod_name),
+                                                      mod.__dict__['__package__'])
+            except Exception:
+                pass  # ignore import exceptions
+
+    @classmethod
+    def _register_commands(cls, module_name):
+        """
+        This method scans all module packages and imports each package module
```

Comment:
This also loads old and stale .pyc files. Can we do it the other way around: Make `CLICommand` register itself in a global static list on initialization?

it also might circumvents existing import statements, which might have some strange side effects. 

Can't we just rely on `CLICommands.COMMAND_LIST` being initialized properly?

Reply:
> If I have some python module `foo` where I have a set of command functions decorated with `@CLICommand`, and this `foo` module is not imported (or transitively imported) from the `module.py` file,

Wouldn't be the simplest solution be to make sure, `foo` is actually transitively imported at module-load time?

Reply:
> Wouldn't be the simplest solution be to make sure, foo is actually transitively imported at module-load time?

I also did that as a first approach, but I end up with import statements in files that are not used and had to add a pylint disable to avoid the `unused-import` error.

> One problem with this approach is: It makes it harder for all MGR module engineers to reason about their modules, just because the dashboard modules contains a module that contains cli commands which is also dynamically loaded.

Can you explain a bit more how this makes the development harder? This code should be transparent for mgr module developers. The only problem I see is if there are global variables in submodules that depend on the initialization order of other modules, but developers should not do those kind of things in the first place. (That's why I'm removing/refactoring two global declarations in the dashboard code in this PR as well)

Reply:
> > Wouldn't be the simplest solution be to make sure, foo is actually transitively imported at module-load time?
> 
> I also did that as a first approach, but I end up with import statements in files that are not used and had to add a pylint disable to avoid the `unused-import` error.

Actually, that sounds like a plan to me! Maybe add a comment about registering cli commands while importing that module?

> 
> > One problem with this approach is: It makes it harder for all MGR module engineers to reason about their modules, just because the dashboard modules contains a module that contains cli commands which is also dynamically loaded.
> 
> Can you explain a bit more how this makes the development harder? This code should be transparent for mgr module developers. The only problem I see is if there are global variables in submodules that depend on the initialization order of other modules

Well, that would already be enough for me to :-1:  this PR :smile: 

> but developers should not do those kind of things in the first place. (That's why I'm removing/refactoring two global declarations in the dashboard code in this PR as well)

Global variables sometimes domake sense. I don't see a reason to break them and prevent everyone from using them.

Reply:
> Actually, that sounds like a plan to me! Maybe add a comment about registering cli commands while importing that module?

Alright, I'll remove the "automatic import" code, and add explicit import of modules where needed.